### DISCUSSION TITLE
Bunit & Csla.Blazor

### DISCUSSION THREAD

**Discussion by ashishsinha24** [MarimerLLC/csla#2341] [Questions]
I am trying to use Bunit to test my Csla Blazor project, but am stuck up in basics.

I am initialising my XUnit constructor with this code.

`  public ControlTest()`
        `{`
            `Services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme).AddCookie();`
            `Services.AddCsla().WithBlazorServerSupport();`

            `CslaConfiguration.Configure().DataPortal().
                DefaultProxy(typeof(Csla.DataPortalClient.HttpProxy), "http://localhost:8040/api/dataportal/");`

            `Csla.ApplicationContext.User = new GenericPrincipal(new GenericIdentity("test"), new[] { "Administrator", "ServiceManager" });`

            `var credentials = Csla.DataPortal.Create<Credentials>("admin", "admin");`
            `var validator = Csla.DataPortal.Fetch<CredentialValidator>(credentials);`
            `var principal = validator.GetPrincipal();`

            `Csla.ApplicationContext.User = principal;`

            `if (principal.Identity.IsAuthenticated)`
            `{`
                `AuthenticationProperties authProperties = new AuthenticationProperties();`
                `App.ApplicationUser = validator;`
            `}`

        `}`

And trying to run this test.

`[Fact]`
        `public async Task TextInputTest()`
        `{`
            `// Arrange`
            `ViewModel<ServiceEdit> vm = new ViewModel<ServiceEdit>();`
            `await vm.RefreshAsync(() => DataPortal.CreateAsync<ServiceEdit>(1));`

            `Services.AddSingleton<ViewModel<ServiceEdit>>(vm);`

            `var component = RenderComponent<TextInput>(`
            `Parameter(nameof(TextInput.Property), vm.GetPropertyInfo("ServiceName")));`

        `}`

But unfortunately I cant even compile this. How do I pass my ViewModel to the test?
Any help would be be nice.

**Comment by rockfordlhotka** [comment]
I've never used bunit.

Normally you need to configure CSLA differently for a unit test framework/runtime than for a normal (aspnet, blazor, winforms) runtime.

In other words, `Services.AddCsla().WithBlazorServerSupport()` is designed to configure CSLA to run inside server-side Blazor, not a unit test runtime.

**Reply by ashishsinha24** [reply]
Thanks,

How can I create a New Blazor ViewModel(). It expects a parameter of type IDataportal<T>.
How can I get it from my businessobject?

**Reply by rockfordlhotka** [reply]
It is designed to use dependency injection, and so in a normal setting it would get that value injected automatically.

When building unit tests you either set up DI, or do it by hand. Either way, you normally inject a mock of the type, rather than the real thing. In other words, you probably don't provide an _actual_ data portal instance, but rather use a mocking tool (Rocks, MOQ, etc.) to create a mock of `IDataPortal` to pass into the constructor.
